### PR TITLE
[FLINK-24717][table-planner]Push down partitions before filters

### PR DESCRIPTION
**Pull request by xuyangzhong**
## What is the purpose of the change

In the optimized period, put the rule PushPartitionsDown before the rule PushFiltersDown to avoid PushFiltersDown consumes all filters and cause that PushFiltersDown will never be applied.

## Brief change log

  - *Change the order between PushPartitionIntoLegacyTableSourceScanRule / PushPartitionIntoTableSourceScanRule and PushFilterIntoTableSourceScanRule / PushFilterIntoLegacyTableSourceScanRule in FlinkStreamRuleSets and FlinkBatchRuleSets.*
  - *Add some test cases to verify this change.*

## Verifying this change

Some test cases are added to verify this change.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented?

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkBatchRuleSets.scala
Comment (KurtYoung) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkBatchRuleSets.scala:182]: how about splitting these rules into 2 sets?

Comment (xuyangzhong) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkBatchRuleSets.scala:182]: My assumption is that the legacy code will be cleaned future, and so it will leave one rule each here. Now maybe to classify them separately is more clear.

**Comment by xuyangzhong** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/PartitionableSourceTest.scala
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/PartitionableSourceTest.scala:134]: Can you add a test case here with the very same query, except you project only `name`? You should see only `name` and `id` in the projected fields of the source. This is the particular case we were hitting.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkBatchRuleSets.scala
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkBatchRuleSets.scala]: introduce constants for the other lists as well to improve readability, in the end this should be only a concatenation of constants instead of this hybrid/nested approach.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/PartitionableSourceTest.scala
Comment (xuyangzhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/PartitionableSourceTest.scala:134]: Just like "select name from PartitionableAndFilterableTable ...(same with now codes)"?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/PartitionableSourceTest.scala
Comment (xuyangzhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/PartitionableSourceTest.scala:66]: I just want to keep consistency with the current code above.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/PartitionableSourceTest.scala
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/PartitionableSourceTest.scala:134]: `select name from PartitionableAndFilterableTable where part1 = 'A' and part2 > 1 and id > 1`

File: flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/TableSourceJsonPlanTest_jsonplan/testPartitionPushDown.out
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/TableSourceJsonPlanTest_jsonplan/testPartitionPushDown.out:64]: is the `toString` consistent now everywhere? first `partitions`, then `filter`, then `project`, then `metadata`. This is how a source should work.

**Review by twalthr — CHANGES_REQUESTED** [review]
I added some minor comment. I will leave the main review up to @slinkydeveloper.

File: flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/TableSourceJsonPlanTest_jsonplan/testPartitionPushDown.out
Comment (xuyangzhong) [reply flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/TableSourceJsonPlanTest_jsonplan/testPartitionPushDown.out:64]: The specs will be print each by the order in which they added to the table source. Now PushDown Partitions will first and partitions info will be print first. I can add a test case about json plan to verify it when the filters and partitions are all exist in the table source.

Comment (twalthr) [reply flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/TableSourceJsonPlanTest_jsonplan/testPartitionPushDown.out:64]: > specs will be print each by the order in which they added to the table source

My question if the print order is consistent everywhere for the specs. But I don't think we need more tests for now. Thanks.

**Review by KurtYoung — CHANGES_REQUESTED** [review]
I think we need to adjust `FlinkBatchProgram` and `FlinkStreamProgram` to split the programs, no?

File: flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/TableSourceJsonPlanTest_jsonplan/testPartitionPushDown.out
Comment (xuyangzhong) [reply flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/TableSourceJsonPlanTest_jsonplan/testPartitionPushDown.out:64]: Sorry, I test some cases and find one bad case when the table source has a watermark. It will miss the rule in optimize period "predicate_pushdown" and match the "logical". So PushDownProject will be called first in "logical", and then PushDownWatermark will be called later in "logical_rewrite", and then PushDownFilter in "logical_rewrite". The source description just like :
`TableSourceScan(table=[[default_catalog, default_database, TestTable, project=[a, b, c], metadata=[], watermark=[c], filter=[>(a, 1)]]], fields=[a, b, c])`
Maybe we need to sort the specs in table source to print them in order if the order is important.

Comment (twalthr) [reply flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/TableSourceJsonPlanTest_jsonplan/testPartitionPushDown.out:64]: Watermarks and filters are a tricky topic anyway, because if the filter is too good, you will have almost no progress in event-time anymore. Let's open a follow up issue for this. I think having a clear spec order could be beneficial.

Comment (xuyangzhong) [reply flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/TableSourceJsonPlanTest_jsonplan/testPartitionPushDown.out:64]: Got it! The issue is here [FLINK-24754].

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkBatchRuleSets.scala
Comment (xuyangzhong) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkBatchRuleSets.scala]: I keep the previous format here.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/PartitionableSourceTest.scala
Comment (xuyangzhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/PartitionableSourceTest.scala:134]: Actually, this case will only remain `name` in the projected field. Because in this case, filters will be pushed down before projection, and projection will not contain the fields  in filters that will no be used later. 
However, if the case contains watermark, filters will be pushed down after projection.And projection fields will contains the fields in filters, just like you said.

**Comment by xuyangzhong** [comment]
@flinkbot run azure

**Comment by xuyangzhong** [comment]
@flinkbot run azure

**Comment by xuyangzhong** [comment]
@flinkbot run azure

**Review by twalthr — APPROVED** [review]
Thanks for the update @xuyangzhong. I'm fine with the changes. @slinkydeveloper also verified them manually. I will merge this later today if there are no objections.

**Comment by xuyangzhong** [comment]
@flinkbot run azure

**Comment by twalthr** [comment]
@xuyangzhong I would like to merge this PR today as it blocks other efforts. It seems there are merge conflicts now. Could you rebase the PR and try to get a SUCCESS build. I will merge asap then.

**Comment by twalthr** [comment]
@xuyangzhong I did the rebase my self to speed things up. Will merge this now, once my build turns green. Thanks.

**Pull request closed without merging**
